apps/aecore/lib/aecore/channel/channel_off_chain_update.ex

```diff
@@ -0,0 +1,164 @@
+
+  @doc """
+  Encodes the given update to a list of binaries.
+  """
+  @spec encode_to_list(update_types()) :: list(binary())
+  def encode_to_list(object) do
+    module = object.__struct__
+    {:ok, tag} = module_to_tag(module)
+    [:binary.encode_unsigned(tag)] ++ module.encode_to_list(object)
+  end
+
+  @doc """
+  Decodes the given update from a list of binaries.
+  """
+  @spec decode_from_list(list(binary())) :: update_types() | error()
+  def decode_from_list([tag | rest]) do
+    decoded_tag = :binary.decode_unsigned(tag)
+
+    case tag_to_module(decoded_tag) do
+      {:ok, module} ->
+        module.decode_from_list(rest)
+
+      {:error, _} = err ->
+        err
+    end
+  end
+
+  @doc """
+  Updates the offchain chainstate acording to the specified update.
+  """
+  @spec update_offchain_chainstate(Chainstate.t() | nil, update_types(), non_neg_integer()) ::
+          {:ok, Chainstate.t()} | error()
+  def update_offchain_chainstate(chainstate, object, channel_reserve) do
+    module = object.__struct__
+    module.update_offchain_chainstate(chainstate, object, channel_reserve)
+  end
+
+  # Function passed to Enum.reduce. Aplies the given update to the chainstate.
+  @spec apply_single_update_to_chainstate(
+          update_types(),
+          {:ok, Chainstate.t() | nil},
+          non_neg_integer()
+        ) :: {:ok, Chainstate.t()} | {:halt, error()}
+  defp apply_single_update_to_chainstate(update, {:ok, chainstate}, channel_reserve) do
+    case update_offchain_chainstate(chainstate, update, channel_reserve) do
+      {:ok, _} = updated_chainstate ->
+        {:cont, updated_chainstate}
+
+      {:error, _} = err ->
+        {:halt, err}
+    end
+  end
+
+  @doc """
+  Applies each update in a list of updates to the offchain chainstate. Breaks on the first encountered error.
+  """
+  @spec apply_updates(Chainstate.t() | nil, list(update_types()), non_neg_integer()) ::
+          {:ok, Chainstate.t()} | error()
+  def apply_updates(chainstate, updates, channel_reserve) do
+    Enum.reduce_while(
+      updates,
+      {:ok, chainstate},
+      &apply_single_update_to_chainstate(&1, &2, channel_reserve)
+    )
+  end
+
+  @doc """
+  Makes sure that for the given account the channel reserve was meet
+  """
+  @spec ensure_channel_reserve_is_met(Account.t(), non_neg_integer()) :: :ok | error()
+  def ensure_channel_reserve_is_met(%Account{balance: balance}, channel_reserve) do
+    if balance < channel_reserve do
+      {:error,
+       "#{__MODULE__} Account does not met channel reserve (We have #{balance} tokens vs channel reserve of #{
+         channel_reserve
+       } tokens)"}
+    else
+      :ok
+    end
+  end
+
+  @doc """
+  Runs preprocess checks for an update which was signed by the foreign peer in the channel.
+  """
+  @spec half_signed_preprocess_check(update_types(), map()) :: :ok | error()
+  def half_signed_preprocess_check(update, opts) do
+    module = update.__struct__
```

Comment:
can you explain this?

Reply:
@thepiwo When one party initializes an update of the channel state and sends an signed transaction to the other party, the other party needs to run some checks to determine if the transaction is not malformed or malicious. All incoming transaction are processed by the same logic in channels which currently looks like this:
1) Run checks determining if the updates in the incoming transaction make sense (transfers are in the right direction, transfer of an positive amount of tokens etc). This step works similar to the validate callback in the on chain transaction validation stack.
2) Check metadata such as signatures, sequence of updates, channel id etc.
3) Run the updates
4) Check if the resulting state hash agrees with the one provided in the transaction
5) Accept the transaction by signing it and mutating the state

This function is an callback for performing the step 1) of the workflow. Maybe this function should have a more self explanatory name (for instance: validate_incoming_halfsigned_transaction)?

Reply:
the "preprocess_check" naming is okay, but what does the update.__struct__ do?

Reply:
@thepiwo Each structure in elixir has a special field called ```__struct__  ``` which contains the module defining the structure. This works exactly like the type field in ```DataTx```.